flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecDeduplicate.java

```diff
@@ -339,25 +362,39 @@ OneInputStreamOperator<RowData, RowData> createDeduplicateOperator() {
                 }
             } else {
                 if (isAsyncStateEnabled()) {
-                    AsyncStateRowTimeDeduplicateFunction processFunction =
-                            new AsyncStateRowTimeDeduplicateFunction(
-                                    rowTypeInfo,
-                                    stateRetentionTime,
-                                    rowtimeIndex,
-                                    generateUpdateBefore,
-                                    generateInsert(),
-                                    keepLastRow);
-                    return new AsyncKeyedProcessOperator<>(processFunction);
+                    if (!keepLastRow && outputInsertOnly) {
+                        checkState(canBeInsertOnly(config, keepLastRow));
+                        // TODO: create async version
+                        return new AsyncKeyedProcessOperator<>(
+                                new RowTimeDeduplicateKeepFirstRowFunction(
+                                        rowTypeInfo, stateRetentionTime, rowtimeIndex));
```

Comment:
Should use the sync version of operator if the function is for sync state:
```suggestion
                        return new KeyedProcessOperator<>(
                                new RowTimeDeduplicateKeepFirstRowFunction(
                                        rowTypeInfo, stateRetentionTime, rowtimeIndex));
```